python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+from temporalio.worker import Worker
+
+@activity.defn
+async def say_hello(name: str) -> str:
+    return f"Hello, {name}!"
+
+@workflow.defn
+class MyWorkflow:
+    @workflow.run
+    async def run(arg1: str) -> str:
+        return await workflow.execute_local_activity(
+            say_hello, arg1, start_to_close_timeout=timedelta(seconds=5)
+        )
+
+
+async def main(stop_event: asyncio.Event):
+    client = await Client.connect("http://localhost:7233")
+    async with Worker(
+        client,
+        task_queue="somequeue",
+        activities=[say_hello],
+        workflows=[MyWorkflow],
+    ):
+        await stop_event.wait()
+```
+
+## Workflow Definitions
+
+A workflow is a class.
+
+Here is an example of a class-based workflow that calls a local activity to get
+a greeting with a signal to update the prefix and a query to get the latest
+greeting:
+
+```python
+import asyncio
+from datetime import timedelta
+from temporalio import activity, workflow
+from temporalio.client import Client
+from temporalio.worker import Worker
+
+@activity.defn
+async def create_greeting_activity(salutation: str, name: str) -> str:
+    return f"{salutation}, {name}!"
+
+@workflow.defn
+class GreetingWorkflow:
+    def __init__() -> None:
+        self._salutation = "Hello"
+        self._salutation_update = asyncio.Event()
+        self._complete = asyncio.Event()
+
+    @workflow.run
+    async def run(self, name: str) -> str:
+        while True:
+            # Store greeting (this is typed)
+            self._current_greeting = await workflow.execute_local_activity(
+                create_greeting_activity,
+                self._salutation,
+                name,
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+            workflow.logger.debug("Greeting set to %s", self._current_greeting)
+            
+            # Wait for salutation update or complete signal (this can be
+            # cancelled)
+            await asyncio.wait(
+                [self._salutation_update.wait(), self._complete.wait()],
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+            if self._salutation_update.is_set():
+                self._salutation_update.clear()
+            elif self._complete.is_set():
+                return self._current_greeting
+
+    @workflow.signal
+    async def update_salutation(self, salutation: str) -> None:
+        self._salutation = salutation
+        self._salutation_update.set()
+
+    @workflow.signal
+    async def complete_with_greeting(self) -> None:
+        self._complete.set()
+
+    @workflow.query
+    async def current_greeting(self) -> str:
+        return self._current_greeting
+
+def test_workflow(client: Client) -> None:
+    # Start the workflow
+    handle = await client.start_workflow(GreetingWorkflow.run, "John")
+
+    # Confirm the current greeting
+    assert "Hello, John!" == await handle.query(GreetingWorkflow.current_greeting)
```

Comment:
What about type safe calls in situations where the caller does not have access to the implementation?
Are you recommending using abstract classes or protocols for those?
Can you add a sample to this proposal?

Reply:
I do not have explicit support for that today and I'm hesitant to do so at least at first.

The protocol would have to match the implementation exactly, down to the proper decorators. But technically it'd work, so you could have:

```python
@workflow.defn
class GreetingWorkflow(Protocol):
    @workflow.run
    async def run(self, name: str) -> str:
        ...

    @workflow.signal
    async def update_salutation(self, salutation: str) -> None:
        ...

    @workflow.signal
    async def complete_with_greeting(self) -> None:
        ...

    @workflow.query
    async def current_greeting(self) -> str:
        ...
```

And then `handle.query(GreetingWorkflow.current_greeting)`. Same with abstract base classes. But all you've really said is that you know how to use common Python interface/inheritance features, it's not specific to Temporal.

My concern is that I am not sure we want to, at least for Python, encourage completely separate pieces of code for the contract. If they want to use inheritance or `@runtime_checkable`, no prob, we support it.

Reply:
👍

Reply:
I have added a new section concerning the separation of contract and impl